### PR TITLE
nvenc, amdvce: rework all user presets/options

### LINKED ISSUE #211: NVIDIA NVENC new preset
### Describe the Bug

Nvidia has made new presets for nvenс. In the next driver / video codec sdk updates old presets will be removed after Q1'2023. 

https://developer.nvidia.com/blog/introducing-video-codec-sdk-10-presets/
https://docs.nvidia.com/video-technologies/video-codec-sdk/deprecation-notices/index.html

### Expected Behavior

Replace old preset to: 
- p1 (perfomance)
- p2
- p3
- p4 (default)
- p5
- p6
- p7 (quality)

Delete old rate control: 
- cbr_ld_hq
- cbr_hq
- vbr_hq

Add tuning info and multi pass settings.

https://docs.nvidia.com/video-technologies/video-codec-sdk/nvenc-preset-migration-guide/index.html

### Additional Context

Performance comparison of all presets
https://github.com/Xaymar/obs-StreamFX/wiki/Encoder-FFmpeg-NVENC#performance-measurements

OBS implementation https://github.com/obsproject/obs-studio/commit/d7c7180a8c55498a484d6e3b132f23e69ab3d553

I compared the old presets with the new ones according to the migration guide.
Used ffmpeg-2022-06-16-git-5242ede48d to encode "Spider-Man Into the Spider-Verse" trailer with settings

- -c:v hevc_nvenc -preset llhq -rc cbr -b:v 20M (identical to current default settings)
- -c:v hevc_nvenc -preset p4 -tune ll -rc cbr -b:v 20M
- -c:v hevc_nvenc -preset p4 -tune ull -rc cbr -b:v 20M

The new preset showed better encode time with slightly lower quality.

And try multi pass encode according to recommended NVENC settings -multipass qres.
https://docs.nvidia.com/video-technologies/video-codec-sdk/nvenc-video-encoder-api-prog-guide/index.html#recommended-nvenc-settings

nvenc

### Sunshine Host Operating System and Version

Windows 10

### Architecture

64 bit

### Sunshine Version

0.12.0

### GPU Type

Nvidia

### GPU Model

GeForce RTX 3060 Ti

### GPU Driver/Mesa Version

512.15

### Capture Method (Linux Only)

_No response_

### PR DESCRIPTION
**Pull request by psyke83**
## Description
**nvenc: rework all user presets/options**
* Presets: remove deprecated presets; add new p1-p7 presets; set "p4" as default.
* Tuning: add new tuning options; set "ull" as default.
* Rate control: remove deprecated options; set "cbr" as default.
* Web UI: keep case consistency (NVEnc -> NVENC, all drop-down options as lowercase);
  ensure all default options are clearly indicated.
* Coder: clarify H264-only status; keep default to "auto".
* Ensure all default presets are correctly identical between C++ and HTML code.
* To improve future compatibility, refer to all possible options via same
  definitions used in ffmpeg code via nvenc headers.
* Docs: update with new profile and tuning options.

**amdvce: rework all user presets/options**
* Quality - fix different values used between codecs; set "balanced" as default.
* Rate control: remove "auto"; set "vbr_latency" as default.
* Web UI: keep consistency (AMD AMF -> AMF, all drop-down options as lowercase);
  ensure all default options are clearly indicated.
* Coder: clarify H264-only status; keep default to "auto".
* Ensure all default presets are correctly identical between C++ and HTML code.
* To improve future compatibility, refer to all possible options via same
  definitions used in ffmpeg code via AMF headers.
* Docs: update with new options and descriptions.

### Screenshot
Screenshot from 2022-12-27 19-50-19

Screenshot from 2022-12-27 19-50-29

### Issues Fixed or Closed
Closes #211

## TODO
- [x] Requires new FFMPEG binaries + NVIDIA headers present for Sunshine to build
- [ ] Verify if build breaks on platforms that don't distribute NVENC headers
- [x] Fix merge conflicts after https://github.com/LizardByte/Sunshine/pull/509 is merged

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [x] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by KuleRucket** [comment]
What is the reason for VBR as default? Nvidia recommends CBR for cloud gaming/streaming. 

https://docs.nvidia.com/video-technologies/video-codec-sdk/nvenc-video-encoder-api-prog-guide/#encoder-tuning-info-and-preset-configurations

File: docs/source/about/advanced_usage.rst [hunk 1]
```diff
@@ -864,28 +887,24 @@ nv_rc
...
+      nv_rc = vbr
 
 nv_coder
 ^^^^^^^^
 
 **Description**
-   The entropy encoding to use.
+   The entropy encoding to use (H264 only).
 
    .. Note:: This option only applies when using nvenc `encoder`_.
```
Comment (ReenigneArcher) [inline docs/source/about/advanced_usage.rst:909]: .. Note:: This option only applies when using H264 with nvenc `encoder`_.

File: docs/source/about/advanced_usage.rst [hunk 2]
```diff
@@ -864,28 +887,24 @@ nv_rc
...
 
-      nv_rc = auto
+      nv_rc = vbr
 
 nv_coder
 ^^^^^^^^
 
 **Description**
-   The entropy encoding to use.
+   The entropy encoding to use (H264 only).
```
Comment (ReenigneArcher) [inline docs/source/about/advanced_usage.rst:907]: Revert and update note.

**Review by ReenigneArcher — COMMENTED** [review]
I just have a minor suggestion to docs

File: docs/source/about/advanced_usage.rst (hunk 2 above)
Comment (psyke83) [reply docs/source/about/advanced_usage.rst]: Should be done now.

**Comment by psyke83** [comment]
@ReenigneArcher:

I've also updated the AMF encoder, as I noticed that the "Quality" setting differs between codecs and therefore requires special handling.

An outstanding issue with this PR is that it may break macos due to the nv/amf headers not being available. Would you prefer that I hardcode the definitions that needs these headers (and thus lose future-proofing compatibility), or exclude the sections via something like `#ifndef __APPLE__`?

If opting for the first option seems best, future-proofing could be improved if we make <libavcodec/nvenc.h> available to Sunshine's build include path as well.

File: docs/source/about/advanced_usage.rst [hunk 3]
```diff
@@ -989,8 +1005,8 @@ amd_coder
    Value      Description
    ========== ===========
    auto       let ffmpeg decide
-   cabac
-   cavlc
+   cabac      context adaptive variable-length coding - higher quality
+   cavlc      context adaptive binary arithmetic coding - faster decode
```
Comment (ReenigneArcher) [inline docs/source/about/advanced_usage.rst:1009]: should this say `faster encode`?

File: docs/source/about/advanced_usage.rst [hunk 4]
```diff
@@ -898,8 +917,8 @@ nv_coder
    Value      Description
    ========== ===========
    auto       let ffmpeg decide
-   cabac
-   cavlc
+   cabac      context adaptive binary arithmetic coding - higher quality
+   cavlc      context adaptive variable-length coding - faster decode
```
Comment (ReenigneArcher) [inline docs/source/about/advanced_usage.rst:921]: should this say `faster encode`?

File: src_assets/common/assets/web/config.html [hunk 5]
```diff
@@ -644,74 +644,72 @@ <h1 class="my-4">Configuration</h1>
...
       <div class="mb-3">
-        <label for="nv_coder" class="form-label">NVEnc Coder</label>
+        <label for="nv_coder" class="form-label">NVENC Coder (H264)</label>
         <select id="nv_coder" class="form-select" v-model="config.nv_coder">
-          <option value="auto">auto</option>
-          <option value="cabac">cabac</option>
-          <option value="cavlc">cavlc</option>
+          <option value="auto">auto -- let ffmpeg decide (default)</option>
+          <option value="cabac">cabac -- context adaptive binary arithmetic coding - higher quality</option>
+          <option value="cavlc">cavlc -- context adaptive variable-length coding - faster decode</option>
```
Comment (ReenigneArcher) [inline src_assets/common/assets/web/config.html:680]: same question

File: src_assets/common/assets/web/config.html [hunk 6]
```diff
@@ -644,74 +644,72 @@ <h1 class="my-4">Configuration</h1>
...
       <div class="mb-3">
-        <label for="amd_coder" class="form-label">AMD AMF Coder</label>
+        <label for="amd_coder" class="form-label">AMF Coder (H264)</label>
         <select id="amd_coder" class="form-select" v-model="config.amd_coder">
-          <option value="auto">auto</option>
-          <option value="cabac">cabac</option>
-          <option value="cavlc">cavlc</option>
+          <option value="auto">auto -- let ffmpeg decide (default)</option>
+          <option value="cabac">cabac -- context adaptive variable-length coding - higher quality</option>
+          <option value="cavlc">cavlc -- context adaptive binary arithmetic coding - faster decode</option>
```
Comment (ReenigneArcher) [inline src_assets/common/assets/web/config.html:712]: same question

File: docs/source/about/advanced_usage.rst (hunk 3 above)
Comment (psyke83) [reply docs/source/about/advanced_usage.rst:1009]: No. The calvc algorithm is less taxing on the decoder side, so can be useful for low power devices that have weak/missing hw decoding.

**Pull request merged**